utils-intervalrdd/src/main/scala/org/bdgenomics/utils/rangearray/RangeSearchableArray.scala

```diff
@@ -0,0 +1,214 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.bdgenomics.utils.rangearray
+
+import com.esotericsoftware.kryo.io.{ Input, Output }
+import com.esotericsoftware.kryo.{ Kryo, Serializer }
+import org.apache.spark.rdd.RDD
+import scala.annotation.tailrec
+import scala.reflect.ClassTag
+
+/**
+ * Companion object for building a RangeSearchableArray from an RDD.
+ */
+object RangeSearchableArray extends Serializable {
+
+  /**
+   * Sorts the RDD and collects it to build RangeSearchableArray, a sorted array that searches
+   * over ranges. This is used for a left side of the broadcast region join in ADAM.
+   *
+   * @param rdd RDD to build a RangeSearchableArray from.
+   * @return The RangeSearchableArray built from this RDD.
+   */
+  def apply[K <: Interval[K]: ClassTag, T: ClassTag](rdd: RDD[(K, T)]): RangeSearchableArray[K, T] = {
+    val sortedArray =
+      rdd.sortByKey()
+        .collect
+
+    new RangeSearchableArray(sortedArray, sorted = true)
+  }
+}
+
+/**
+ * Originally, a RangeSearchableArray was a collection of trees.
+ * Alas, we have no trees anymore.
+ * I blame global warming.
+ *
+ * @param arr An array of values for the left side of the join. We require
+ *   this array to be sorted.
+ */
+case class RangeSearchableArray[K <: Interval[K], T: ClassTag](arr: Array[(K, T)], sorted: Boolean = false) {
+
+  // ensure that array is sorted
+  val array =
+    if (sorted) arr
+    else arr.sortBy(_._1)
+
+  def length = array.length
+  def midpoint = pow2ceil()
+
+  @tailrec private def pow2ceil(i: Int = 1): Int = {
+    if (2 * i >= length) {
+      i
+    } else {
+      pow2ceil(2 * i)
+    }
+  }
+
+  @tailrec private def binarySearch(rr: K,
+                                    idx: Int = 0,
+                                    step: Int = midpoint): Option[Int] = {
+    if (array.length == 0) {
+      None
+    } else if (rr.overlaps(array(idx)._1)) {
+      Some(idx)
+    } else if (step == 0) {
+      None
+    } else {
+      val stepIdx = idx + step
+      val nextIdx = if (stepIdx >= length ||
+        (!rr.overlaps(array(stepIdx)._1) &&
+          rr.compareTo(array(stepIdx)._1) < 0)) {
+        idx
+      } else {
+        stepIdx
+      }
+      binarySearch(rr, nextIdx, step / 2)
+    }
+  }
+
+  @tailrec private def expand(rr: K,
+                              idx: Int,
+                              step: Int,
+                              list: List[(K, T)] = List.empty): List[(K, T)] = {
+    if (idx < 0 ||
+      idx >= length ||
+      !rr.overlaps(array(idx)._1)) {
```

Comment:
Yeah, there's a bit more to the fix. I've got a patch locally (with the proof) that resolves this. I'm just putting the finishing on the proof, will push shortly.